enhancements/installer/component-selection.md

```diff
@@ -0,0 +1,319 @@
+---
+title: component-selection-during-install
+authors:
+  - "@bparees"
+reviewers:
+  - install team - need agreement on install-config api updates and CVO config rendering
+  - cvo team - need agreement on CVO resource filtering api and new behavior
+approvers:
+  - @decarr - support for configurable CVO-managed content set feature
+creation-date: 2021-05-04
+last-updated: 2021-05-04
+status: provisional
+---
+
+# User Selectable Install Solutions
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement proposes a mechanism for cluster installers to exclude one or more optional components for
+their installation which will determine which payload components are/are not installed in their cluster.
+Core components are defined as the set of Second Level Operators managed by the Cluster Version Operator
+which today cannot be disabled until after completing the install and editing a CVO override, or editing
+the CVO overrides as part of rendering+editing manifests.
+
+The proposed UX is to make this a first class part of the install config api with the implementation
+being arguments supplied to the CVO to filter the user-selected manifests.
+
+## Motivation
+
+There is an increasing desire to move away from "one size fits all" cluster installations, and
+towards flexibility about what should/should not exist in a new cluster out of the box.  This can
+be seen in efforts such as hypershift, single node, and code-ready-containers.  Each of these
+efforts has done some amount of one-off work to enable their requirements.  This EP proposes a
+mechanism that allows components to be disabled in a first class way that the installer exposes.
+
+### Goals
+
+* Admins can easily explicitly exclude specific "optional" components from their cluster, at install time.
+* Admins can enable a previously excluded optional component, at runtime.
+
+### Non-Goals
+
+* Making control-plane critical components optional (k8s apiserver, openshift apiserver, openshift controller,
+  networking, etc)
+* Defining which components should be disable-able (this will be up to component teams to classify themselves
+as `addons` or not)
+* Providing a way to install OLM operators as part of the initial cluster install.  This EP is about making
+the install experience around the existing CVO-based components more flexible, not adding new components to the
+install experience.
+* Allowing components to be disabled post-install.
```

Comment:
```suggestion
* Allowing components to be disabled post-install.
```
```suggestion
* Allowing components to be disabled post-install.
* Defining disabled components via any means other than the ExcludeAddons list (ie: no defaulting to disable a second level operator based on platform.
```
Maybe this is implicit in the the next item but it seemed worth calling out.

Reply:
> @staebler similar to comments you made below. Do you feel like this is worth calling out?

I think there is value in calling that out explicitly. I don't see it being implied by the next item, beyond one particular way of disabling components.

Reply:
> i would say there's no reason the installer couldn't pre-configure the install-config to exclude certain components based on the target platform

The installer cannot easily support defaults for the exclusions, because the installer cannot distinguish between an empty field and an omitted field.

Reply:
I would like to explicitly state that the Installer is not responsible for maintaining configuration matrices that affect component exclusion. Either you get the default components installed or you specify those you wish to exclude.

Reply:
I think @dhellmann was envisioning defaulting behavior like "don't enable cluster baremetal operator if we're installing on AWS" which isn't a great fit for cluster profiles (how many cluster profiles do we need to create to cover all the permutations of "with" and "without" baremetal, along w/ other dimensions that might be toggled on or off?)

but i'm not necessarily committed to the idea that the installer does anything smart here, this EP enables someone to build an install-config generator that makes smart choices, it doesn't mean the installer has to be responsible for making those smart choices.

But it sounds like something that we need to discuss w/ @dhellmann to get agreement on.